### ISSUE TITLE
Jittery Playback in Windowed mode in Surface

### ISSUE THREAD

**Issue by primaeval** [koying/SPMC#615]
The jittery windowed playback bug is back. SPMC was working perfectly whereas Kodi was jittery.
Now they are both as bad as each other. You win some, you lose some. ;)
Here is an example of the play method using windowed mode in TV Guide Fullscreen.
https://github.com/primaeval/script.tvguide.fullscreen/blob/799bfb3808e0a2c47b8b3f3d2a2442c09a40ffe8/gui.py#L2117
You can recreate the error by playing the BBC iPlayer streams from plugin.video.iplayerwww in the main repo in a TV Guide such as TV Guide Fullscreen.

**Comment by koying** [comment]
Debug log?
As a reminder, there is a builtin log uploader at "Settings-System-Logging-Upload a debug log..."

**Comment by primaeval** [comment]
Thanks. This problem keeps coming back. It is better on some Android devices than others. The Nvidia Shield has been pretty good so far until 16.5.x.

Bad 16.5.3 http://paste.semperpax.com/pvj3ndfmg
Good 16.4.0 https://gist.github.com/primaeval/efa7a5031aa697af9edc643051da803b

Using TV Guide Fullscreen and BBC TV addons from my repo:
https://github.com/primaeval/repository.primaeval

**Comment by koying** [comment]
The obvious difference is that 16.5.x uses Mediacodec(Surface) by default again, as kodi
What if you disable it?

**Comment by primaeval** [comment]
Well spotted. That seems to fix it. Thanks and Merry Christmas.

Perhaps it would be better turned off by default again. Several people that have told me they also had to downgrade to 16.4 because of this.

**Comment by koying** [comment]
Is there a pattern re system having issues ?

**Comment by primaeval** [comment]
The only pattern I could see was your stuff was good, Kodi was bad. ;)
But there is something very strange about starting video's in windowed mode.
There might not be so much optimisation then as whoever wrote the code thought that in windowed mode the video would always be smaller than fullscreen and so the quality didn't matter so much.
The TV Guides use windowed mode in a window that is as big as the desktop because it allows for osd overlays.

**Comment by JFG90** [comment]
I'm also getting jittery playback when playing live tv streams, never used to get it before on 16.4.2 is there anyway this can be fixed?

**Comment by koying** [comment]
@primaeval Could you test http://download.semperpax.com/spmc/android-arm/SPMC-spmc-jarvis-72302b2-armeabi-v7a.apk

Should fix the jitter in Windowed Surface mode

**Comment by primaeval** [comment]
Excellent. That works perfectly. Thanks. What did you do?

**Comment by koying** [comment]
Err... Fixed Windowed mode with Surface ;)

If you want the gorry details, windowed mode + multi-layered video (as Surface is, vs. the "normal" way, which is OpenGL) was designed for aml and rpi, which use a somewhat different rendering loop.

**Comment by primaeval** [comment]
Thanks. I saw a ton of changes in the last commit to do with surface. 
It's still holiday mode in my brain so I was just looking for the executive summary.

**Comment by JFG90** [comment]
I've been getting jittery playback on tv streams so hopefully this should fix?

**Comment by koying** [comment]
Ah, yeah. I pushed unrelated code regarding dynamic allocation of surfaces.

@JonnyGadd Yeah, the build should fix that as well, although it's an unrelated issue

**Comment by JFG90** [comment]
Awesome thanks I'll test it and get back to you, this version ok to run on the shield?

**Comment by primaeval** [comment]
It ran on my shield tv box perfectly.

**Comment by JFG90** [comment]
Just tested and my playback still becomes jittery after 30 seconds or so it's odd it never used to do it on 16.4.2

**Comment by primaeval** [comment]
Here is the debug log from my test for a minute or so of CBBC in TV Guide Fullscreen.
It worked fine.
http://paste.semperpax.com/pty5p3cgv

**Comment by koying** [comment]
@JonnyGadd Please push a log: https://github.com/koying/SPMC/wiki/How-do-I-publish-a-debug-log%3F

**Comment by JFG90** [comment]
Ok I will do

**Comment by JFG90** [comment]
Log uploaded 

pyt9wizly

**Comment by primaeval** [comment]
@JonnyGadd 16.5-3 doesn't work. Try the link koying posted to 16.6.0
http://download.semperpax.com/spmc/android-arm/SPMC-spmc-jarvis-72302b2-armeabi-v7a.apk

**Comment by koying** [comment]
@JonnyGadd Besides, you seem to have another issue:

> 11:49:53.006 T:1488927024   ERROR: SQL: [Textures13.db] Some kind of disk I/O error occurred

You might want to delete Textures13.db so that it's re-created

**Comment by JFG90** [comment]
I had tried on 16.6 and it was the same so reverted back, ok delete textures, where would I find that? I made my own build so would it delete the skin?

**Comment by koying** [comment]
The db is in /sdcard/Android/data/com.semperpax.spmc16/files/.spmc/userdata/Database/

**Comment by JFG90** [comment]
Ok cool thankyou so hopefully it will fix the picture jittering?

**Comment by koying** [comment]
Not a clue, but I'll definitely not investigate any issues on a borked install ;)

**Comment by primaeval** [comment]
@JonnyGadd I would recommend renaming the whole spmc data folder so you can isolate this jittering issue before you start fixing everything else.
rename
/sdcard/Android/data/com.semperpax.spmc16
to 
/sdcard/Android/data/com.semperpax.spmc16.save

**Comment by JFG90** [comment]
Fair enough I understand I'm guessing you mean a build is a borked install? Lol

**Comment by koying** [comment]
Not necessarily, although it makes debugging harder, I just mean that an install with a corrupted database is not a good base to investigate ;)

**Comment by JFG90** [comment]
Ok if you like if i still get issues I will back my build up on a fresh install of spmc no build?

**Comment by JFG90** [comment]
Primaeval what would renaming it do?

**Comment by JFG90** [comment]
Can't find a textures.db

**Comment by primaeval** [comment]
In this case nothing  I think, because koying has made a separate folder for the test so it won't overwrite your main settings. I've got a lot of renamed folders in data now so please correct me if I'm wrong.
Have a look for com.semperpax.spmctest and com.semperpax.spmc16.
/sdcard/Android/data/com.semperpax.spmctest

**Comment by JFG90** [comment]
I don't have that in mine mate, just what the spmc folder is usually named

**Comment by primaeval** [comment]
The textures db path should be something like this then:
/sdcard/Android/data/com.semperpax.spmc16/files/.spmc/userdata/Database/Textures13.db

Make sure you have hidden files turned on in a decent File Manager.
The .spmc folder will be hidden by default.

**Comment by JFG90** [comment]
Yes mate use es file Explorer and it's not there :(

**Comment by primaeval** [comment]
Just the textures db or the whole com.semperpax.spmc16 tree?

**Comment by JFG90** [comment]
Textures.db isn't in the database folder mate

**Comment by primaeval** [comment]
Even when you restart SPMC?

**Comment by JFG90** [comment]
You mean restart then minimise spmc and then have look?

**Comment by primaeval** [comment]
Yes. If you look in your log there is something very strange going on with textures13.db
```
ERROR: SQL: [Textures13.db] Some kind of disk I/O error occurred
                                            Query: SELECT id, cachedurl, lasthashcheck, imagehash, width, height FROM texture JOIN sizes ON (texture.id=sizes.idtexture AND sizes.size=1) WHERE url='/storage/emulated/0/Android/data/com.semperpax.spmc16/files/.spmc/addons/script.exodus.artwork/resources/media/exodus/next.png'
```
That is why you need to isolate this jittering problem from the exodus problem that is giving you a disk I/O error.

**Comment by JFG90** [comment]
It's not exodus that's giving me the jittery playback it's flawless IPTV streams, they never used to jitter on 16.4.2 only since these new updates

**Comment by primaeval** [comment]
Have you installed 16.6.0 yet?

**Comment by JFG90** [comment]
Yer tried it earlier and it was still the same I'm out now so will have to try look for texture file again later

**Comment by primaeval** [comment]
We're going around in circles here.
1. exit spmc
2. rename /sdcard/Android/data/com.semperpax.spmc16 to /sdcard/Android/data/com.semperpax.spmc16.save
3. install http://download.semperpax.com/spmc/android-arm/SPMC-spmc-jarvis-72302b2-armeabi-v7a.apk
4. start spmc 16.6.0
5. install your tv guide
6. install your iptv addon
7. play a jittery stream in your tv guide
8. post the log here

**Comment by JFG90** [comment]
Ah right ok i get you now yes alright I'll do that :) won't be til tomorrow now though, I'll keep posted thanks for your help!

**Comment by JFG90** [comment]
Guys just thought I'd report back, so it turns out the jittering I was getting must have been server related from the IPTV service, as they were down for maintenance this morning and after testing when it was all back up and running it seems fine, 16.6 be pushed on the playstore?

**Comment by JFG90** [comment]
So it turns out it's still jittering, could someone have a look I've uploaded a new log, on 16.6 fresh install wih no build, paste is po8qfs8at

**Comment by primaeval** [comment]
From that log it looks like you played a channel from the flawless addon directly, not in windowed mode from a TV Guide. If that's the case and it jitters it that seems like a different issue to the one in this Issue that koying fixed.
I'll butt out now and let the master take over. :)

**Comment by JFG90** [comment]
Yes i didn't think it would matter doing it from the guide as it's the same issues using directly through the addon, will he see this?

**Comment by JFG90** [comment]
Can you see an issue looking at the log then?

**Comment by primaeval** [comment]
Try turning debug on so that there is more information for koying to look at in the log.
If you can, try it with the BBC iplayerwww live streams in a tv guide because those streams are more stable and should isolate the issue better.

**Comment by JFG90** [comment]
Streams in the tv guide do the exact same issue, will make another log later with de bug on

**Comment by primaeval** [comment]
I found a strange but probably related issue to this.
A 10bit video file would play fine if an overlay was enabled but only the audio would play without the overlay.
I made another Issue here:
https://github.com/koying/SPMC/issues/685
